Source/WebKit/UIProcess/API/Swift/WebPage.swift

```diff
@@ -21,34 +21,136 @@
+        set { backingWebView.isInspectable = newValue }
+    }
+
     private let backingNavigationDelegate: WKNavigationDelegateAdapter
 
+    @ObservationIgnored
+    private var observations = KeyValueObservations()
+
+    @ObservationIgnored
     private lazy var backingWebView: WKWebView = {
         let webView = WKWebView(frame: .zero)
         webView.navigationDelegate = backingNavigationDelegate
         return webView
     }()
 
     public init() {
+        // FIXME: Consider whether we want to have a single value here or if the getter for `navigations` should return a fresh sequence every time.
         let (stream, continuation) = AsyncStream.makeStream(of: NavigationEvent.self)
-        self.navigations = Navigations(source: stream)
-        self.backingNavigationDelegate = WKNavigationDelegateAdapter(navigationProgressContinuation: continuation)
+        navigations = Navigations(source: stream)
+
+        backingNavigationDelegate = WKNavigationDelegateAdapter(navigationProgressContinuation: continuation)
+
+        observations.contents = [
+            createObservation(for: \.url, backedBy: \.url),
+            createObservation(for: \.title, backedBy: \.title),
+            createObservation(for: \.estimatedProgress, backedBy: \.estimatedProgress),
+            createObservation(for: \.isLoading, backedBy: \.isLoading),
+            createObservation(for: \.serverTrust, backedBy: \.serverTrust),
+            createObservation(for: \.hasOnlySecureContent, backedBy: \.hasOnlySecureContent),
+            createObservation(for: \.themeColor, backedBy: \.themeColor),
+        ]
+    }
+
+    @discardableResult
+    public func load(_ request: URLRequest) -> NavigationID? {
+        backingWebView.load(request).map(NavigationID.init(_:))
     }
 
-    @discardableResult public func load(_ request: URLRequest) -> NavigationID? {
-        self.backingWebView.load(request).map(NavigationID.init(_:))
+    @discardableResult
+    public func load(htmlString: String, baseURL: URL) -> NavigationID? {
+        backingWebView.loadHTMLString(htmlString, baseURL: baseURL).map(NavigationID.init(_:))
     }
 
-    @discardableResult public func load(htmlString: String, baseURL: URL) -> NavigationID? {
-        self.backingWebView.loadHTMLString(htmlString, baseURL: baseURL).map(NavigationID.init(_:))
+    private func createObservation<Value, BackingValue>(for keyPath: KeyPath<WebPage_v0, Value>, backedBy backingKeyPath: KeyPath<WKWebView, BackingValue>) -> NSKeyValueObservation {
+        backingWebView.observe(backingKeyPath, options: [.prior, .old, .new]) { [_$observationRegistrar, unowned self] _, change in
+            if change.isPrior {
+                _$observationRegistrar.willSet(self, keyPath: keyPath)
+            } else {
+                _$observationRegistrar.didSet(self, keyPath: keyPath)
+            }
+        }
+    }
+}
+
+extension WebPage_v0 {
+    private struct KeyValueObservations: ~Copyable {
```

Comment:
I’m curious why you’re using this wrapper type instead of just storing these directly on `WebPage` and removing them in `WebPage.deinit`?